net/queue/ff.go

```diff
@@ -0,0 +1,256 @@
+package queue
+
+import (
+	"context"
+	"hash/fnv"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/textileio/go-threads/core/thread"
+)
+
+type linkedOperation struct {
+	prev, next *linkedOperation
+	tid        thread.ID
+	call       PeerCall
+	priority   int
+	created    int64
+}
+
+type peerQueue struct {
+	index       map[thread.ID]*linkedOperation
+	first, last *linkedOperation
+	sync.Mutex
+}
+
+// Simple FIFO-queue with O(1)-operations.
+func newPeerQueue() *peerQueue {
+	return &peerQueue{index: make(map[thread.ID]*linkedOperation)}
+}
+
+// Add new call to the queue or replace existing one with lower priority.
+func (q *peerQueue) Add(tid thread.ID, call PeerCall, priority int) bool {
+	op, exist := q.index[tid]
+	if !exist {
+		// append new entry at the end
+		op = &linkedOperation{
+			tid:      tid,
+			call:     call,
+			priority: priority,
+			created:  time.Now().Unix(),
+		}
+		if q.last == nil {
+			// empty queue
+			q.first = op
+			q.last = op
+		} else {
+			q.last.next = op
+			op.prev = q.last
+			q.last = op
+		}
+		q.index[tid] = op
+		return true
+	}
+
+	if op.priority < priority {
+		// just replace the call
+		op.call = call
+	}
+	return false
+}
+
+// Return previously added calls in FIFO order.
+func (q *peerQueue) Pop() (PeerCall, thread.ID, int64, bool) {
+	if q.first == nil {
+		return nil, thread.Undef, 0, false
+	}
+	op := q.first
+
+	q.first = op.next
+	if q.first != nil {
+		q.first.prev = nil
+	} else {
+		q.last = nil
+	}
+
+	delete(q.index, op.tid)
+	return op.call, op.tid, op.created, true
+}
+
+// Remove corresponding call if it was scheduled.
+func (q *peerQueue) Remove(tid thread.ID) bool {
+	op, exist := q.index[tid]
+	if !exist {
+		return false
+	}
+
+	switch {
+	case q.last == op && q.first == op:
+		// single operation - empty the queue
+		q.first = nil
+		q.last = nil
+	case q.first == op:
+		// first operation
+		next := op.next
+		next.prev = nil
+		q.first = next
+	case q.last == op:
+		// last operation
+		prev := op.prev
+		prev.next = nil
+		q.last = prev
+	default:
+		prev, next := op.prev, op.next
+		prev.next = next
+		next.prev = prev
+	}
+	delete(q.index, tid)
+	return true
+}
+
+func (q *peerQueue) Size() int {
+	return len(q.index)
+}
+
+/* Call queue implementation */
+
+var _ CallQueue = (*ffQueue)(nil)
+
+type ffQueue struct {
```

Comment:
receiving, keeping and executing calls evenly and in time for the multiple peers